pkg/controller/bundlec/hash.go

```diff
@@ -0,0 +1,116 @@
+package bundlec
+
+import (
+	"hash"
+	"sort"
+
+	"github.com/pkg/errors"
+	core_v1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/util/sets"
+)
+
+func (st *resourceSyncTask) hashSecretRef(name, namespace string, filter sets.String, optional *bool, hash hash.Hash) error {
+	secret, exists, err := st.derefObject(core_v1.SchemeGroupVersion.WithKind("Secret"), name, namespace)
+	if err != nil {
+		return err
+	}
+	if !exists {
+		if optional == nil || !*optional {
+			return errors.Errorf("secret %q not found", name)
+		}
+		return nil
+	}
+
+	found, err := hashSecret(secret.(*core_v1.Secret), hash, filter)
+	if err != nil {
+		return err
+	}
+	if !found && (optional == nil || !*optional) {
+		return errors.Errorf("not all keys %v found in secret %q", filter, name)
+	}
+
+	return nil
+}
+
+func (st *resourceSyncTask) hashConfigMapRef(name, namespace string, filter sets.String, optional *bool, hash hash.Hash) error {
+	configmap, exists, err := st.derefObject(core_v1.SchemeGroupVersion.WithKind("ConfigMap"), name, namespace)
+	if err != nil {
+		return err
+	}
+	if !exists {
+		if optional == nil || !*optional {
+			return errors.Errorf("configMap %q not found", name)
+		}
+		return nil
+	}
+
+	found, err := hashConfigMap(configmap.(*core_v1.ConfigMap), hash, filter)
+	if err != nil {
+		return err
+	}
+	if !found && (optional == nil || !*optional) {
+		return errors.Errorf("not all keys %v found in configmap %q", filter, name)
+	}
+
+	return nil
+}
+
+// hashConfigMap hashes the sorted values in the configmap in sorted order
+// with a NUL as a separator character between and within pairs of key + value.
+func hashConfigMap(configMap *core_v1.ConfigMap, h hash.Hash, filter sets.String) (bool, error) {
```

Comment:
You could also keep the hash type and ignore the errors. It is documented to never return errors. Less boilerplate. Or just use `utilruntime.Must()` as an error sink to make sure nothing weird happens unnoticed.

Reply:
I've opted to remain using hash.Hash since the point of this method is to pump things into a hash for later hashing, but I cast to io.Writer to enable the use of io.WriteString